TaskWithSmartPointers/MySharedPtr.h

```diff
@@ -0,0 +1,128 @@
+#pragma once
+#include <iostream>
+// #include <atomic>
+#include <mutex> 
+
+template <class T>
+class MySharedPtr
+{
+public:
+	std::mutex mtx;
+
+	T* get()
+	{
+		return m_ptr;
+	}
+
+	long use_count() const noexcept
+	{
+		return m_referencesCounter ? *m_referencesCounter : 0;
+	}
+
+	bool unique() const noexcept
+	{
+		return (1 == use_count());
+	}
+
+	void reset(T* p)
+	{
+		if (m_referencesCounter != nullptr)
+		{
+			mtx.lock();
+			(*m_referencesCounter) -= 1;
+			if (*m_referencesCounter == 0)
+			{
+				delete m_ptr;
+				delete m_referencesCounter;
+			}
+			mtx.unlock();
+		}
+		m_ptr = p;
+		m_referencesCounter = new long(1);
+	}
+
+	T& operator*()
+	{
+		return *(this->m_ptr);
+	}
+
+	MySharedPtr() // default constructor
+	{
+		m_ptr;
+		m_referencesCounter;
+	}
+
+	MySharedPtr(T* ptr)
+	{
+		m_ptr = ptr;
+		m_referencesCounter = new long(1);
+	}
+
+	MySharedPtr(const MySharedPtr& obj) // copy constructor 
+	{
+		m_ptr = obj.m_ptr;
+		m_referencesCounter = obj.m_referencesCounter;
+		mtx.lock();
+		if (obj.m_ptr != nullptr)
+		{
+			(*m_referencesCounter)++;
+		}
+		mtx.unlock();
+	}
+
+	MySharedPtr<T>& operator=(MySharedPtr<T>& obj) // copy assignment
+	{
+		m_ptr = obj.m_ptr;
+		m_referencesCounter = obj.m_referencesCounter;
+		mtx.lock();
+		if (obj.m_ptr != nullptr)
```

Comment:
if (obj.m_referenceCounter != nullptr)?

Reply:
This condition does not need to be checked

Reply:
My concern was there's dereferencing of the m_referenceCounter within the 'if' clause. I understand that whenever we have non-null m_ptr we care m_referenceCounter is created. But the code feels more consistent if we check the pointer we're going to dereference is not nullptr.

Reply:
Done

Reply:
Please keep only the 'if' clause with 'm_referenceCounter' and remove with  the 'm_ptr'. Shouldn't be a big effort to do.